src/lib.rs

```diff
@@ -1,14 +1,207 @@
+            A,
+            B,
+            C,
+        }
+
+        struct GroupedOrderingFoo {
+            #[allow(dead_code)]
+            groups: [GroupedOrderingFooGroup; 3],
+            index_lookup: HashMap<GroupedOrderingFooGroup, usize>,
+        }
+
+        impl GroupedOrdering for GroupedOrderingFoo {
+            type Group = GroupedOrderingFooGroup;
+
+            fn compare(
+                &self,
+                a: &GroupedOrderingFooGroup,
+                b: &GroupedOrderingFooGroup,
+            ) -> Ordering {
+                self.index_lookup[a].cmp(&self.index_lookup[b])
+            }
+        }
+
+        impl TryFrom<[GroupedOrderingFooGroup; 3]> for GroupedOrderingFoo {
+            type Error = String;
+
+            fn try_from(groups: [GroupedOrderingFooGroup; 3]) -> Result<Self, Self::Error> {
+                let index_lookup: HashMap<GroupedOrderingFooGroup, usize> = groups
+                    .into_iter()
+                    .enumerate()
+                    .map(|(index, group)| (group, index))
+                    .collect();
+                if index_lookup.len() < 3 {
+                    return Err("Found duplicate groups".to_owned());
+                }
+                Ok(Self {
+                    groups,
+                    index_lookup,
+                })
+            }
+        }
+
+        impl GroupedOrderable<GroupedOrderingFoo> for u32 {
+            fn map_to_grouped_ordering(&self) -> GroupedOrderingFooGroup {
+                match self % 3 {
+                    0 => GroupedOrderingFooGroup::A,
+                    1 => GroupedOrderingFooGroup::B,
+                    2 => GroupedOrderingFooGroup::C,
+                    _ => unreachable!(),
+                }
+            }
+        }
+
+        let grouped_ordering: GroupedOrderingFoo = [
+            GroupedOrderingFooGroup::A,
+            GroupedOrderingFooGroup::B,
+            GroupedOrderingFooGroup::C,
+        ]
+        .try_into()
+        .unwrap();
+
+        let mut nums: Vec<u32> = vec![0, 1, 2, 3, 4, 5];
+        nums.sort_by_grouped_ordering(&grouped_ordering);
+        assert_that!(&nums).is_equal_to(vec![0, 3, 1, 4, 2, 5]);
+    }
 
     #[test]
-    fn it_works() {
-        let result = add(2, 2);
-        assert_eq!(result, 4);
+    fn test_vec_sort() {
+        grouped_ordering!(GroupedOrderingFoo, [A, B, C]);
+
+        impl GroupedOrderable<GroupedOrderingFoo> for u32 {
+            fn map_to_grouped_ordering(&self) -> GroupedOrderingFooGroup {
+                match self % 3 {
+                    0 => GroupedOrderingFooGroup::A,
+                    1 => GroupedOrderingFooGroup::B,
+                    2 => GroupedOrderingFooGroup::C,
+                    _ => unreachable!(),
+                }
+            }
+        }
+
+        let grouped_ordering: GroupedOrderingFoo = [
+            GroupedOrderingFooGroup::A,
+            GroupedOrderingFooGroup::B,
+            GroupedOrderingFooGroup::C,
+        ]
```

Comment:
It is very mildly tempting to generate another macro eg here `grouped_ordering_foo![B, A, C]` to reduce the boilerplate of grouped-ordering-specification instantiation but that would probably be stupid overkill (though maybe fun to implement, I have the feeling I might have to use that "you can't generate a new proc macro from a proc macro but you can generate a declarative macro that "forwards" its arguments to another proc macro" technique that I described in that "fully-populated collection" blog post)

Reply:
UPDATE: did this in #2